Add a template project

matt-oakes opened the pull request:
This PR adds a project template which uses [Turborepo](https://turborepo.com) and [Vite](https://vite.dev] as the foundation. It then has a built-in "generator" which allows generation of COGS plugins and custom content either with or without React.

This will allow developers to very easily get started with the COGS SDK without requiring them to set up the project structure.

I have updated the main README to include details of using the template. The previous "manual" instructions are retained, but are now hidden by default to show only the recommended route and make the tips section at the bottom of the README easier to spot.

To test this before the PR is merged in, you can include the branch name when generating from the template with:

```bash
npx create-turbo@latest --example https://github.com/clockwork-dog/cogs-sdk/tree/template/template --package-manager yarn
```

chetbox commented on template/.github/workflows/test.yml:
```diff
@@ -0,0 +1,71 @@
+      - run: pnpm run build
+      # Copy the test-project-pack with symlinks followed and copied
+      - run: cp -rL test-project-pack test-project-pack-no-symlinks
+      # Upload a zip artifact for this run
+      - uses: actions/upload-artifact@v4
+        with:
+          name: ${{ github.event.repository.name }}-${{ github.sha }}
+          path: test-project-pack-no-symlinks
```
It would be great to also upload zips to a release if this is a release build. This is our usual pattern.

matt-oakes replied:
Done!

chetbox commented on template/packages/types/src/Utils.ts:
```diff
@@ -0,0 +1,5 @@
+export type DeepPartial<T> = T extends object
```
This doesn't appear to be used anywhere. Is there any reason to keep it in the template?

chetbox commented on template/packages/typescript-config/vite.json:
```diff
@@ -0,0 +1,8 @@
+  "extends": "./base.json",
+  "compilerOptions": {
+    "lib": ["ESNext", "DOM"],
+    "useDefineForClassFields": true
```
Why is this included here?

It looks like the default depends on your `target` which seems sensible at first glance.
https://www.typescriptlang.org/tsconfig/#useDefineForClassFields

chetbox commented on template/pnpm-lock.yaml:
```diff
@@ -0,0 +1,5862 @@
+lockfileVersion: '9.0'
```
Does it make sense to check in the lock file? I would imagine this doesn't do anything if you pick a different package manager anyway?

chetbox commented on template/prettier.config.mjs:
```diff
@@ -0,0 +1,3 @@
+export default {
+  plugins: ["prettier-plugin-organize-imports"],
```
❤️

chetbox commented on template/pnpm-workspace.yaml:
```diff
@@ -0,0 +1,9 @@
+packages:
```
Should this pnpm-specific file be included?

chetbox replied:
Do we also need to add something similar for Yarn in that case?  It's worth keeping in mind we generally use Yarn internally.

matt-oakes replied:
Turbo automatically adds the config needed for Yarn or NPM if you choose those package managers.

We still need to include this pnpm config file as it adds config to tell it which Node.JS packages should have their `postinstall` scripts run to build or download the native part of the module. It will just not be used can can be deleted if using a different package manager.

chetbox commented on template/turbo/generators/config.ts:
```diff
@@ -0,0 +1,148 @@
+import { mkdir, symlink } from "node:fs/promises";
+import path from "node:path";
+
+export default function generator(plop: PlopTypes.NodePlopAPI): void {
```
I'll be honest, I don't understand this file. 😅

chetbox commented on template/turbo/generators/templates-shared/eslint.config.ts.hbs:
```diff
@@ -0,0 +1,10 @@
+import tseslint from "typescript-eslint";
+
+export default tseslint.config([
+  viteEslintConfig,{{#ifCond templateType '===' "react"}}
+  reactEslintConfig,{{/ifCond}}
```
🤯 Mad but cool.

matt-oakes commented:
> Weird. I've not had this with Vite before. I'll update.

Vite v7 came out a week or so ago, so it's likely a new requirement with that version.

matt-oakes commented on template/turbo/generators/package.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "type": "commonjs"
+}
```
Turborepo generators need to be CommonJS otherwise they don't work correctly. This file was generated by Turborepo itself. The documentation for generator is here with the note about ESM:

https://turborepo.com/docs/guides/generating-code#custom-generators

matt-oakes commented on template/packages/typescript-config/package.json:
```diff
@@ -0,0 +1,11 @@
+  "name": "@repo/typescript-config",
+  "version": "0.0.0",
+  "private": true,
+  "peerDependencies": {
+    "@types/node": "^22.0.0"
+  },
+  "devDependencies": {
+    "@types/node": "^22.16.0"
+  }
```
`peerDependency` means anything that installs this module needs to supply a `@types/node` which matches our requirements.

`devDependency` means to use that version of `@types/node` when we're running dev tools like TypeScript and ESLint inside that package.

In this case the versions are the same, but quite often you'd have a looser version specification for `peerDependencies` so users of your package can select a version which works for them and then a more specific version in `devDependencies` when you're running your own dev tools.

matt-oakes commented on template/turbo/generators/templates-shared/package.json.hbs:
```diff
@@ -0,0 +1,33 @@
+    "lint": "eslint"
+  },
+  "dependencies": {
+    "@clockworkdog/cogs-client": "^2.8.1"{{#ifCond templateType '===' "react"}},
+    "@clockworkdog/cogs-client-react": "^2.8.1",
```
Hmmmm... yes we do actually, but without a lockfile we'd only need to update this when a major version changes due to the `^` at the start.

Personally, this is why I like tagged releases to have the version number commited into the repo rather than having `0.0.0` in the modules and only setting and releasing them in Github Actions. If we changed the release process to commit the version numbers in the `package.json` we could make the release script also update the template version number in a `postversion` script.

matt-oakes replied:
Fixed.

matt-oakes commented:
Added a new commit with some of the suggestions you made.

chetbox commented on README.md:
````diff
@@ -46,7 +46,28 @@ A COGS plugin is a folder in the `plugins` directory of a COGS Project. The plug
+
+You can then look at the README of the generated project to generate custom content or a plugin.
+
+```sh
+npx turbo generate
+yarn turbo generate
+pnpm turbo generate
+```
````
I find it's handy to put these in separate code blocks because GitHub gives you a handy copy-to-clipboard button. There's not much point copying them all in one go!

chetbox commented on template/README.md:
```diff
@@ -0,0 +1,79 @@
+
+This will prompt you to select the type of content or plugin you want to create. Follow the prompts to create your new content or plugin.
+
+Once this has been created make sure the depenencies are installed with `pnpm install` and then you can build or start in dev mode.
```
> depenencies

typo

chetbox commented:
I had a problem getting started with a JS plugin with `yarn`.

![yarn-plugin-error rec](https://github.com/user-attachments/assets/7efb6e84-cd39-43ff-a3af-a8e5bd965a84)

matt-oakes commented:
Fixed the comments above. I have made the template Yarn 4 by default and added a note to the README to say other package managers are possible, but you need to manualy set them up after using the template as `create-turbo` doesn't correctly move between the different package managers.

chetbox commented:
I wonder if we should make these options  a little clearer.

<img width="768" height="265" alt="image" src="https://github.com/user-attachments/assets/85409f29-e399-43cf-9c95-2f3b2e8eb70b" />



e.g. 

- Media Master custom content
- COGS Plugin

chetbox commented:
I used `yarn turbo generate` to add a project.

Then `yarn ; yarn start` gives me an error:

<img width="1946" height="1030" alt="image" src="https://github.com/user-attachments/assets/86ac6a71-d46b-4cb5-acdc-7c9afa385df9" />

matt-oakes commented:
You didn't run `yarn install` after generating.

In both README file I have made the command you copy and run this to avoid this issue:

```sh
yarn turbo generate && yarn install
```

If you ran it like that `yarn start` would have worked right away.

Let me know if it needs to be made clearer.

matt-oakes commented:
We could do, but I was trying to avoid too much "magic" and just spell out what needs to be done with Turborepo directly.

I can add that `new` script in if you want though.

chetbox commented:
I've tested creating a plugin, custom content, JS and JS + React.

The developer workflow is really nice. Great work, Matt!

I really like how plugin/Media Master simulator in dev mode redirects so you can just use the dev tools in COGS. 👏🏼 

> I can add that `new` script in if you want though.

Just this one small thing and it's good to go. 3️⃣ 2️⃣ 1️⃣  🚀 🌔
